File: mobile_app/lib/services/supabase_service.dart
import 'dart:async';
import 'package:supabase_flutter/supabase_flutter.dart';
import 'package:mobile_app/config/config.dart';
import 'package:mobile_app/models/models.dart';

/// Exception for Supabase operations
class SupabaseException implements Exception {
  final String message;
  SupabaseException(this.message);

  @override
  String toString() => message;
}

/// Service for interacting with Supabase database
/// Handles:
/// - Fetching schemes
/// - Saving user profiles
/// - Bookmarking schemes
/// - Analytics tracking
class SupabaseService {
  late final SupabaseClient _supabase;

  SupabaseService() {
    _supabase = Supabase.instance.client;
  }

  // ========== LOGGING ==========
  void _log(String level, String message) {
    print('[$level] $message');
  }

  // ========== INITIALIZATION ==========

  /// Initialize Supabase connection
  /// Call this during app startup
  static Future<void> initialize() async {
    try {
      await Supabase.initialize(
        url: AppConfig.supabaseUrl,
        anonKey: AppConfig.supabaseAnonKey,
      );
      print('[INFO] ✅ Supabase initialized successfully');
    } catch (e) {
      print('[ERROR] ❌ Supabase initialization failed: $e');
      rethrow;
    }
  }

  // ========== SCHEME QUERIES ==========

  /// Fetch all schemes from database
  /// Returns list of 823 schemes
  Future<List<Scheme>> fetchAllSchemes() async {
    try {
      _log('INFO', '🔄 Fetching all schemes from database...');

      final response = await _supabase
          .from(AppConfig.schemesTable)
          .select()
          .timeout(AppConfig.apiTimeout);

      final schemes = (response as List)
          .map((json) => Scheme.fromJson(json as Map<String, dynamic>))
          .toList();

      _log('INFO', '✅ Fetched ${schemes.length} schemes');
      return schemes;
    } catch (e) {
      final error = 'Error fetching schemes: $e';
      _log('ERROR', error);
      throw SupabaseException(error);
    }
  }

  /// Fetch schemes filtered by category
  Future<List<Scheme>> fetchSchemesByCategory(String category) async {
    try {
      _log('DEBUG', '🔄 Fetching schemes for category: $category');

      final response = await _supabase
          .from(AppConfig.schemesTable)
          .select()
          .eq('category_name', category)
          .timeout(AppConfig.apiTimeout);

      final schemes = (response as List)
          .map((json) => Scheme.fromJson(json as Map<String, dynamic>))
          .toList();

      _log('INFO', '✅ Fetched ${schemes.length} schemes for category: $category');
      return schemes;
    } catch (e) {
      _log('ERROR', '❌ Error fetching schemes by category: $e');
      rethrow;
    }
  }

  /// Fetch schemes for specific state (central + state-specific)
  Future<List<Scheme>> fetchSchemesByState(String state) async {
    try {
      _log('DEBUG', '🔄 Fetching schemes for state: $state');

      final response = await _supabase
          .from(AppConfig.schemesTable)
          .select()
          .or('is_central.eq.true,applicable_states.contains."$state"')
          .timeout(AppConfig.apiTimeout);

      final schemes = (response as List)
          .map((json) => Scheme.fromJson(json as Map<String, dynamic>))
          .toList();

      _log('INFO', '✅ Fetched ${schemes.length} state schemes for: $state');
      return schemes;
    } catch (e) {
      _log('ERROR', '❌ Error fetching schemes by state: $e');
      rethrow;
    }
  }

  /// Search schemes by keyword
  Future<List<Scheme>> searchSchemes(String query) async {
    try {
      _log('DEBUG', '🔄 Searching schemes with query: $query');

      final lowerQuery = query.toLowerCase();

      final response = await _supabase
          .from(AppConfig.schemesTable)
          .select()
          .or('title.ilike.%$lowerQuery%,description.ilike.%$lowerQuery%,benefits.ilike.%$lowerQuery%')
          .timeout(AppConfig.apiTimeout);

      final schemes = (response as List)
          .map((json) => Scheme.fromJson(json as Map<String, dynamic>))
          .toList();

      _log('INFO', '✅ Found ${schemes.length} schemes matching: "$query"');
      return schemes;
    } catch (e) {
      _log('ERROR', '❌ Error searching schemes: $e');
      rethrow;
    }
  }

  /// Fetch new schemes added this month
  Future<List<Scheme>> fetchNewSchemes() async {
    try {
      _log('DEBUG', '🔄 Fetching new schemes...');

      final response = await _supabase
          .from(AppConfig.schemesTable)
          .select()
          .eq('is_new', true)
          .order('created_at', ascending: false)
          .limit(50)
          .timeout(AppConfig.apiTimeout);

      final schemes = (response as List)
          .map((json) => Scheme.fromJson(json as Map<String, dynamic>))
          .toList();

      _log('INFO', '✅ Fetched ${schemes.length} new schemes');
      return schemes;
    } catch (e) {
      _log('ERROR', '❌ Error fetching new schemes: $e');
      rethrow;
    }
  }

  // ========== BOOKMARKS / SAVED SCHEMES ==========

  /// Save scheme to user's bookmarks
  Future<void> bookmarkScheme({
    required String userUuid,
    required String schemeId,
  }) async {
    try {
      _log('DEBUG', '🔄 Bookmarking scheme $schemeId for user $userUuid');

      await _supabase
          .from(AppConfig.savedSchemesTable)
          .insert({
            'user_uuid': userUuid,
            'scheme_id': schemeId,
            'created_at': DateTime.now().toIso8601String(),
          })
          .timeout(AppConfig.apiTimeout);

      _log('INFO', '✅ Scheme bookmarked successfully');
    } on PostgrestException catch (e) {
      if (e.code == '23505') {
        _log('INFO', '⚠️ Scheme already bookmarked');
      } else {
        _log('ERROR', '❌ Database error: ${e.message}');
        rethrow;
      }
    } catch (e) {
      _log('ERROR', '❌ Error bookmarking scheme: $e');
      rethrow;
    }
  }

  /// Remove scheme from bookmarks
  Future<void> unbookmarkScheme({
    required String userUuid,
    required String schemeId,
  }) async {
    try {
      _log('DEBUG', '🔄 Removing bookmark for scheme $schemeId');

      await _supabase
          .from(AppConfig.savedSchemesTable)
          .delete()
          .eq('user_uuid', userUuid)
          .eq('scheme_id', schemeId)
          .timeout(AppConfig.apiTimeout);

      _log('INFO', '✅ Scheme unbookmarked successfully');
    } catch (e) {
      _log('ERROR', '❌ Error unbookmarking scheme: $e');
      rethrow;
    }
  }

  /// Get all bookmarked schemes for user
  Future<List<String>> getBookmarkedSchemeIds(String userUuid) async {
    try {
      _log('DEBUG', '🔄 Fetching bookmarked schemes for user $userUuid');

      final response = await _supabase
          .from(AppConfig.savedSchemesTable)
          .select('scheme_id')
          .eq('user_uuid', userUuid)
          .timeout(AppConfig.apiTimeout);

      final schemeIds = (response as List)
          .map((item) => item['scheme_id'] as String)
          .toList();

      _log('INFO', '✅ Fetched ${schemeIds.length} bookmarked schemes');
      return schemeIds;
    } catch (e) {
      _log('ERROR', '❌ Error fetching bookmarked schemes: $e');
      rethrow;
    }
  }

  // ========== ANALYTICS & TRACKING ==========

  /// Record scheme view (for analytics)
  Future<void> recordSchemeView({
    required String userUuid,
    required String schemeId,
  }) async {
    try {
      if (!AppConfig.enableAnalytics) return;

      _log('DEBUG', '📊 Recording scheme view');

      // Increment view count in schemes table
      await _supabase
          .from(AppConfig.schemesTable)
          .update({'view_count': 'view_count + 1'})
          .eq('id', schemeId)
          .timeout(AppConfig.apiTimeout);

      _log('DEBUG', '✅ View recorded');
    } catch (e) {
      _log('INFO', '⚠️ Error recording view (non-critical): $e');
      // Don't rethrow - this is non-critical
    }
  }

  /// Save user profile to analytics table
  Future<void> saveUserProfileAnalytics(UserProfile profile) async {
    try {
      if (!AppConfig.enableAnalytics) return;

      _log('DEBUG', '📊 Saving user profile to analytics');

      await _supabase
          .from(AppConfig.userAnalyticsTable)
          .insert({
            'user_uuid': profile.uuid,
            'gender': profile.gender,
            'age': profile.age,
            'state': profile.state,
            'occupation': profile.occupation,
            'created_at': DateTime.now().toIso8601String(),
          })
          .timeout(AppConfig.apiTimeout);

      _log('INFO', '✅ User profile saved to analytics');
    } on PostgrestException catch (e) {
      if (e.code == '23505') {
        _log('INFO', '⚠️ User profile already exists in analytics');
      } else {
        _log('ERROR', '❌ Database error: ${e.message}');
      }
    } catch (e) {
      _log('INFO', '⚠️ Error saving analytics (non-critical): $e');
      // Don't rethrow - this is non-critical
    }
  }

  // ========== CONNECTION MANAGEMENT ==========

  /// Check if connected to Supabase
  bool isConnected() {
    try {
      return _supabase.auth.currentUser != null ||
          _supabase.auth.currentSession == null;
    } catch (e) {
      return false;
    }
  }

  /// Close Supabase connection
  Future<void> disconnect() async {
    try {
      _log('DEBUG', '🔄 Disconnecting from Supabase...');
      // Supabase doesn't require explicit disconnect in Flutter
      _log('INFO', '✅ Supabase connection closed');
    } catch (e) {
      _log('ERROR', '❌ Error disconnecting: $e');
    }
  }
}

File: backend/app/services/supabase_service.py
"""Supabase service for database operations"""

import os
from datetime import datetime
from typing import List, Optional
from supabase import create_client
from app.models.scheme import Scheme
from app.models.user import User
from config.settings import settings


class SupabaseService:
    """Service for Supabase database operations"""
    
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(SupabaseService, cls).__new__(cls)
            cls._instance._supabase = None
        return cls._instance
    
    @property
    def client(self):
        """Get or create Supabase client"""
        if self._supabase is None:
            self._supabase = create_client(
                settings.SUPABASE_URL,
                settings.SUPABASE_SERVICE_KEY
            )
        return self._supabase
    
    # ========== SCHEMES ==========
    
    def get_all_schemes(self) -> List[Scheme]:
        """Fetch all schemes from database"""
        try:
            response = self.client.table('schemes').select('*').execute()
            schemes = [Scheme(**item) for item in response.data]
            return schemes
        except Exception as e:
            raise Exception(f"Error fetching schemes: {str(e)}")
    
    def search_schemes(
        self, 
        query: str, 
        state: Optional[str] = None,
        category: Optional[str] = None,
        limit: int = 50
    ) -> List[Scheme]:
        """Search schemes by query, state, or category"""
        try:
            db_query = self.client.table('schemes').select('*')
            
            # Add filters
            if state:
                db_query = db_query.eq('state_name', state)
            if category:
                db_query = db_query.eq('category_name', category)
            
            # Search in title or description
            if query:
                # Use Supabase full-text search
                db_query = db_query.or_(
                    f"title.ilike.%{query}%,description.ilike.%{query}%,benefits.ilike.%{query}%"
                )
            
            response = db_query.limit(limit).execute()
            schemes = [Scheme(**item) for item in response.data]
            return schemes
        except Exception as e:
            raise Exception(f"Error searching schemes: {str(e)}")
    
    def get_schemes_by_state(self, state: str) -> List[Scheme]:
        """Get schemes for specific state"""
        try:
            # Central schemes + state-specific schemes
            response = self.client.table('schemes').select('*').or_(
                f"state_name.eq.{state},is_central.eq.true"
            ).execute()
            schemes = [Scheme(**item) for item in response.data]
            return schemes
        except Exception as e:
            raise Exception(f"Error fetching schemes by state: {str(e)}")
    
    def get_scheme_by_id(self, scheme_id: str) -> Optional[Scheme]:
        """Get single scheme by ID"""
        try:
            response = self.client.table('schemes').select('*').eq(
                'id', scheme_id
            ).single().execute()
            return Scheme(**response.data) if response.data else None
        except Exception as e:
            raise Exception(f"Error fetching scheme: {str(e)}")
    
    # ========== USERS ==========
    
    def get_user_profile(self, user_uuid: str) -> Optional[User]:
        """Get user profile"""
        try:
            response = self.client.table('users').select('*').eq(
                'uuid', user_uuid
            ).single().execute()
            return User(**response.data) if response.data else None
        except Exception:
            # User might not exist yet
            return None
    
    def create_or_update_user(
        self,
        user_uuid: str,
        gender: Optional[str] = None,
        age: Optional[int] = None,
        state: Optional[str] = None,
        occupation: Optional[str] = None
    ) -> User:
        """Create or update user profile"""
        try:
            user_data = {
                "uuid": user_uuid,
                "gender": gender,
                "age": age,
                "state": state,
                "occupation": occupation,
                "updated_at": datetime.utcnow().isoformat()
            }
            
            # Try to insert, if exists update
            response = self.client.table('users').upsert(user_data).execute()
            return User(**response.data[0]) if response.data else User(**user_data)
        except Exception as e:
            raise Exception(f"Error updating user: {str(e)}")
    
    # ========== BOOKMARKS ==========
    
    def bookmark_scheme(self, user_uuid: str, scheme_id: str) -> bool:
        """Save scheme for user"""
        try:
            bookmark_data = {
                "user_uuid": user_uuid,
                "scheme_id": scheme_id,
                "saved_at": datetime.utcnow().isoformat()
            }
            self.client.table('saved_schemes').upsert(bookmark_data).execute()
            return True
        except Exception as e:
            raise Exception(f"Error bookmarking scheme: {str(e)}")
    
    def remove_bookmark(self, user_uuid: str, scheme_id: str) -> bool:
        """Remove bookmarked scheme"""
        try:
            self.client.table('saved_schemes').delete().eq(
                'user_uuid', user_uuid
            ).eq('scheme_id', scheme_id).execute()
            return True
        except Exception as e:
            raise Exception(f"Error removing bookmark: {str(e)}")
    
    def get_saved_schemes(self, user_uuid: str) -> List[str]:
        """Get list of saved scheme IDs for user"""
        try:
            response = self.client.table('saved_schemes').select(
                'scheme_id'
            ).eq('user_uuid', user_uuid).execute()
            return [item['scheme_id'] for item in response.data]
        except Exception as e:
            raise Exception(f"Error fetching saved schemes: {str(e)}")


# Create singleton instance
supabase_service = SupabaseService()
